static_frame/core/index_hierarchy.py

```diff
@@ -456,211 +719,273 @@ def _from_type_blocks(cls: tp.Type[IH],
+        self._map = HierarchicalLocMap(indices=self._indices, indexers=self._indexers)
 
-        if self._levels.depth <= 1:
-            raise ErrorInitIndex(f'invalid depth ({self._levels.depth}) for IndexLevels composed in IndexHierarchy')
+    def _update_array_cache(self: IH) -> None:
+        # This MUST be set before entering this context
+        assert self._pending_extensions is not None
 
-        self._recache = self._blocks is None
-        self._name = None if name is NAME_DEFAULT else name_filter(name)
+        new_indexers = [np.empty(self.__len__(), DTYPE_INT_DEFAULT) for _ in range(self.depth)]
 
-    #---------------------------------------------------------------------------
-    def __deepcopy__(self: IH, memo: tp.Dict[int, tp.Any]) -> IH:
+        current_size = len(self._blocks)
+
+        for depth, indexer in enumerate(self._indexers):
+            new_indexers[depth][:current_size] = indexer
+
+        # TODO-Ariza: Should we clear self._indexers?
+        # TODO-Ariza: What if an error is raised anywhere onward in this method?
+
+        offset = current_size
+
+        # For all these extensions, we have already update self._indices - we now need to map indexers
+        for pending in self._pending_extensions: # pylint: disable = E1133
+            if isinstance(pending, PendingRow):
+                for depth, label_at_depth in enumerate(pending):
+                    label_index = self._indices[depth]._loc_to_iloc(label_at_depth)
+                    new_indexers[depth][offset] = label_index
+
+                offset += 1
+            else:
+                group_size = len(pending)
+
+                for depth, (self_index, other_index) in enumerate(
+                        zip(self._indices, pending._indices) # type: ignore
+                    ):
+                    remapped_indexers_unordered = other_index._index_iloc_map(self_index)
+                    remapped_indexers_ordered = remapped_indexers_unordered[
+                        pending._indexers[depth] # type: ignore
+                    ]
+
+                    new_indexers[depth][offset:offset + group_size] = remapped_indexers_ordered
+
+                offset += group_size
+
+        self._pending_extensions.clear()
+        self._indexers = np.array(new_indexers)
+        self._indexers.flags.writeable = False
+        self._blocks = self._create_blocks_from_self()
+        self._values = self._blocks.values
+        self._map = HierarchicalLocMap(indices=self._indices, indexers=self._indexers)
+        self._recache = False
+
+    # --------------------------------------------------------------------------
+
+    def __setstate__(self, state: tp.Tuple[None, tp.Dict[str, tp.Any]]) -> None:
+        '''
+        Ensure that reanimated NP arrays are set not writeable.
+        '''
+        for key, value in state[1].items():
+            setattr(self, key, value)
+        self._values.flags.writeable = False
```

Comment:
Does `writeable` need to be set for `self._indexers`?

Reply:
I've made `_indexers` non-writeable everywhere. I'm not necessarily concerned about someone doing it, but I feel like it's better to ensure state doesn't change wherever possible

Reply:
Makes sense.